cmd/kubeadm/app/cmd/join.go

```diff
@@ -122,35 +106,38 @@ func NewJoin(cfgPath string, args []string, cfg *kubeadmapi.NodeConfiguration, s
 		fmt.Println("Skipping pre-flight checks")
 	}
 
-	ok, err := kubeadmutil.UseGivenTokenIfValid(&cfg.Secrets)
-	if !ok {
-		if err != nil {
-			return nil, fmt.Errorf("%v (see --help)\n", err)
-		}
-		return nil, fmt.Errorf("Must specify --token (see --help)\n")
-	}
-
 	return &Join{cfg: cfg}, nil
 }
 
 // Run executes worked node provisioning and tries to join an existing cluster.
 func (j *Join) Run(out io.Writer) error {
-	clusterInfo, err := kubenode.RetrieveTrustedClusterInfo(j.cfg)
+	clusterInfo, err := kubenode.RetrieveTrustedClusterInfo(j.cfg.Discovery.Token)
 	if err != nil {
 		return err
 	}
 
-	connectionDetails, err := kubenode.EstablishMasterConnection(j.cfg, clusterInfo)
-	if err != nil {
-		return err
-	}
-
-	kubeconfig, err := kubenode.PerformTLSBootstrap(connectionDetails)
-	if err != nil {
-		return err
+	var cfg *clientcmdapi.Config
+	// TODO: delete this first block when we move Token to the discovery interface
```

Comment:
It's a todo to move token discovery behind the discovery interface witch involved deleteing ConnectionInfo. I don't want to do that here.